inheritdoc order fixes #54494

ZacharyPatten opened the pull request:
This PR add ordering to the logic of processing `inheritdoc`s to ensure that explicit XML on members overrides XML from `inheritdoc`s. fixes #54494

Coinsidering this is my first PR to https://github.com/dotnet/roslyn, I'm opening this as a draft PR until it is reviewed.

dnfadmin commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/dotnet/roslyn?pullRequest=54618) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/roslyn?pullRequest=54618) before we can accept your contribution.<br/><br/>:x: ZacharyPatten [sign now](https://cla.dotnetfoundation.org/dotnet/roslyn?pullRequest=54618)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/roslyn?pullRequest=54618) it.</sub>

ZacharyPatten commented:
@sharwell Can has a code review please?

sharwell reviewed (changes requested):
Can you add a test for the following?

```cs
/// <summary>hello world</summary>
public class A { }

/// <summary>
/// <inheritdoc/>
/// <para>goodbye world</para>
/// </summary>
public class B : A { }
```

ZacharyPatten commented:
Good catch. :thumbsup:

As you predicted, the tests you recommended failed. We only want to reorder top level `iheritdoc`s, not inline ones. I updated the code so that instead of only `RewriteMany`, there is now `RewriteManyInline` and `RewriteManyTopLevel` so that we can selectively apply the re-ordering logic to top level `inheritdoc`s only. I believe this fixed the issue. Both of the tests you recommended have been added and I added a third _more extreme_ test case as well.

I believe it is ready for another review.

ZacharyPatten commented:
Wait... Sorry I want to check on one more thing... I want to add another test for inheritdoc chaining.

ZacharyPatten commented:
So... `inheritdoc` doesn't play well with `<param>`. `<para>` is intended to be used for formatting, not really be considered a logical XML node, but because it is technically an XML node, putting `inheritdoc` inside of a `<param>` will pull (and concatenate all content inside of the `<param>` nodes of the `inheritdoc` source). Example:
```cs
/// <summary>
/// aaa
/// <para>bbb</para>
/// <para>ccc</para>
/// </summary>
class A { }

/// <summary>
/// ddd
/// <para><inheritdoc /></para>
/// </summary>
class B : A { }
```
I personally find that very unintuitive. I would have expected the Quick Info Content on `B` to be:
```
ddd
aaa
bbb
ccc
```
but the content is not that. It is:
```
ddd
bbbccc
```
However unintuitive that is... I assume that is fully intentional. And regardless, that is a separate issue from this PR.

That being said... I added another unit test, and I was worried for nothing. The unit test passed. So It is ready for a code review again. :)

ZacharyPatten commented:
This example (along with the `<para>` example previously mentioned), makes me wonder if there aren't a lot more bugs related to `inheritdoc` that need to be flushed out. Are explicit XML docs supposed to hide inherited ones?

```cs
/// <summary>a</summary>
/// <remarks>A</remarks>
class A { }

/// <summary>b</summary>
/// <inheritdoc />
class B : A { }

/// <summary><inheritdoc /></summary>
/// <remarks>C</remarks>
class C : B { }
```

The quick info on C shows "ba" instead of just "b". The explicit `<summary>` on `B` is not hiding the summary on `A`.

One way around this is to rewrite the code as follows:

```cs
/// <summary>a</summary>
/// <remarks>A</remarks>
class A { }

/// <summary>b</summary>
/// <remarks><inheritdoc /></remarks>
class B : A { }

/// <summary><inheritdoc /></summary>
/// <remarks>C</remarks>
class C : B { }
```

but I just wonder if `inheritdoc` needs to only work on top-level nodes and only allow inheritance to pull unique nodes (only allow one node of each node type)

sharwell commented:
> ... putting `inheritdoc` inside of a `<para>` ...

This portion of inheritdoc was never implemented. In this case, there is a failure to update the path to include a `para[0]` filter. There is no need to add special support for inheritdoc inside block-level elements in this pull request.

> Are explicit XML docs supposed to hide inherited ones?
> ...
> The quick info on C shows "ba" instead of just "b".

So this is a fun one:

* Due to the way inheritdoc is evaluated, `B` has two `<summary>` elements. Quick Info only shows the first one.
* The `path` query created for evaluation of the inline `inheritdoc` element in `C` specifies `/summary` instead of `/summary[0]`, which causes it to concatenate the contents of _both_ `<summary>` elements in `B`.

This isn't the desired behavior but it isn't something introduced by this pull request so we can leave it for a separate bug.

ZacharyPatten commented:
That makes a lot more sense now. I draft-ed the PR because when I saw examples like those I got really confused and I was worried that this PR was not the appropriate way to handle this topic.

You obviously have much more experience with this topic than I do (I've only been seriously looking into `inheritdoc` in the last week really). If you feel like this PR was the appropriate way to handle the issue, I will un-draft it. :)

ZacharyPatten commented:
By the way... one more funny, _niche_ edge case I just want to mention... :D

```cs
/// <summary>a</summary>
/// <inheritdoc>hello world</inheritdoc>
class A { }

/// <summary><inheritdoc><inheritdoc /></inheritdoc></summary>
class B : A { }
```

It seems like the quick info on `B` should be  `ahelloworld` but instead it is the `<summary>` from `System.Object`, `Supports all classes in the .NET class hierarchy and provides low-level services to derived classes. This is the ultimate base class of all .NET classes; it is the root of the type hierarchy.`
